src/client/datascience/jupyter/kernels/kernelConnectionWrapper.ts

```diff
@@ -147,49 +187,55 @@ export class KernelConnectionWrapper implements Kernel.IKernelConnection {
     ): Kernel.IControlFuture<IDebugRequestMsg, IDebugReplyMsg> {
-        return this.kernelConnection.requestDebug(content, disposeOnDone);
+        return this.getKernelConnection().requestDebug(content, disposeOnDone);
     }
     requestIsComplete(content: { code: string }): Promise<IIsCompleteReplyMsg> {
-        return this.kernelConnection.requestIsComplete(content);
+        return this.getKernelConnection().requestIsComplete(content);
     }
     requestCommInfo(content: { target_name?: string | undefined }): Promise<ICommInfoReplyMsg> {
-        return this.kernelConnection.requestCommInfo(content);
+        return this.getKernelConnection().requestCommInfo(content);
     }
     sendInputReply(content: IReplyErrorContent | IReplyAbortContent | IInputReply): void {
-        return this.kernelConnection.sendInputReply(content);
+        return this.getKernelConnection().sendInputReply(content);
     }
     createComm(targetName: string, commId?: string): Kernel.IComm {
-        return this.kernelConnection.createComm(targetName, commId);
+        return this.getKernelConnection().createComm(targetName, commId);
     }
     hasComm(commId: string): boolean {
-        return this.kernelConnection.hasComm(commId);
+        return this.getKernelConnection().hasComm(commId);
     }
     registerCommTarget(
         targetName: string,
         callback: (comm: Kernel.IComm, msg: ICommOpenMsg<'iopub' | 'shell'>) => void | PromiseLike<void>
     ): void {
-        return this.kernelConnection.registerCommTarget(targetName, callback);
+        return this.getKernelConnection().registerCommTarget(targetName, callback);
     }
     removeCommTarget(
         targetName: string,
         callback: (comm: Kernel.IComm, msg: ICommOpenMsg<'iopub' | 'shell'>) => void | PromiseLike<void>
     ): void {
-        return this.kernelConnection.removeCommTarget(targetName, callback);
+        return this.getKernelConnection().removeCommTarget(targetName, callback);
     }
     registerMessageHook(
         msgId: string,
         hook: (msg: IIOPubMessage<IOPubMessageType>) => boolean | PromiseLike<boolean>
     ): void {
-        return this.kernelConnection.registerMessageHook(msgId, hook);
+        return this.getKernelConnection().registerMessageHook(msgId, hook);
     }
     removeMessageHook(
         msgId: string,
         hook: (msg: IIOPubMessage<IOPubMessageType>) => boolean | PromiseLike<boolean>
     ): void {
-        return this.kernelConnection.removeMessageHook(msgId, hook);
-    }
-    shutdown(): Promise<void> {
-        return this.kernel.dispose();
+        return this.getKernelConnection().removeMessageHook(msgId, hook);
+    }
+    async shutdown(): Promise<void> {
+        if (
+            this.kernel.kernelConnectionMetadata.kind === 'startUsingRemoteKernelSpec' ||
+            this.kernel.kernelConnectionMetadata.kind === 'connectToLiveKernel'
+        ) {
+            await this.kernel.session?.shutdown();
```

Comment:
When shutting down a kernel, we must explicitly call `session.shutdown` for remote kernels.
As `shutdown` means kill kernel & in our code base `IKernel.dispose` will never shutdown the remote kernels.

Reply:
This change ensures 3rd party extensions get the desired effect as calling shutdown should ideally shutdown the kernel whether its local or remote.

Reply:
Doesn't shutdown also dispose? Seems possible we don't need the kind check here but can just call shutdown? 

This shutdown is an explicit shutdown from the other extension so seems like calling shutdown explicitly would make sense?

Reply:
> Doesn't shutdown also dispose?

Yes it does. however dispose doesn't shutdown.

> This shutdown is an explicit shutdown from the other extension so seems like calling shutdown explicitly would make sense?

Yes, thats the reason i'm ensuring shutdown will kill the kernel.
In our code we had shutdown which would NOT kill the kernel & we never used that code path, we always called dispose (which would shutdown the kernel if its a local kernel).

Reply:
Yes so couldn't this code just call ```this.kernel.session?.shutdown()``` every time? Instead of the condition?

Reply:
You're making me look like an idiot now, 
yes absolutely right. no idea why i even wrote this code.
lol...

Reply:
Aah yes, now i remember, that's because calling `kernel.dispose` will clean everything properly.
So `kernel.dispose` is absolutely necessary (i.e. cleanup).

Reply:
We could optionally call shutdown but that won't be a clean way, as that would kill the session and our code could see that as an unhandled (premature) termination of the kernel.